webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,234 @@
+import { Button, Col, InputNumber, Modal, Row } from 'antd';
+import { Form, Input, Select } from 'antd';
+import { ModalProps } from 'antd/es/modal/Modal';
+import yaml from 'js-yaml';
+import React, { useCallback, useEffect, useState } from 'react';
+import MonacoEditor from 'react-monaco-editor';
+
+import useStorage from 'hooks/useStorage';
+import { getResourcePools, getTemplates } from 'services/api';
+import { NotebookConfig, RawJson, ResourcePool, Template } from 'types';
+import { launchNotebook, previewNotebook } from 'utils/task';
+
+import Link from './Link';
+import RadioGroup from './RadioGroup';
+
+const { Option } = Select;
+const { Item } = Form;
+
+const STORAGE_PATH = 'notebook-launch';
+const STORAGE_KEY = 'notebook-config';
+
+interface Props extends ModalProps {
+  onLaunch?: () => void;
+  visible?: boolean;
+}
+
+const NotebookModal: React.FC<Props> = (
+  { visible = false, onLaunch = (() => null), ...props }: Props,
```

Comment:
**suggestion:** my preference is to leave it undefined (not setting a default), and wherever you are running `onLaunch` to do `if (onLaunch) onLaunch();`. The idea is that checking whether `onLaunch` is defined or not is a cheaper operation than having JS to schedule the `onLaunch` function call into the event loop and creating a call stack for it, even though it simply returns a `null`.